File: sdk/typescript/src/evolution-runtime.ts
/**
 * EvolutionRuntime — High-level evolution API for agents.
 *
 * Composes EvolutionCache + SignalEnrichment + async outbox into two simple methods:
 *   - suggest(error, context?) → strategy recommendation (<1ms local, fallback to server)
 *   - learned(error, outcome, summary, geneId?) → fire-and-forget outcome recording
 *
 * Also handles:
 *   - Bootstrap: auto-load sync snapshot on init
 *   - Periodic sync: pull delta every N seconds
 *   - Session tracking: correlates suggest → learned within a task
 *
 * Usage:
 *   const runtime = new EvolutionRuntime(client.im.evolution);
 *   await runtime.start();
 *
 *   const fix = await runtime.suggest('ETIMEDOUT: connection timed out');
 *   // ... agent applies fix.strategy ...
 *   runtime.learned('ETIMEDOUT', 'success', 'Fixed by increasing timeout to 30s');
 */

import { EvolutionCache } from './evolution-cache';
import { extractSignals, createEnrichedExtractor } from './signal-enrichment';
import type { SignalTag, ExecutionContext, SignalEnrichmentConfig, IMGene } from './types';

/** Minimal interface for the evolution client — avoids circular import */
interface EvolutionClientLike {
  getSyncSnapshot(since?: number): Promise<{ data?: any }>;
  analyze(options: Record<string, any>): Promise<{ data?: any }>;
  record(options: Record<string, any>): Promise<{ data?: any }>;
  sync(options: Record<string, any>): Promise<{ data?: any }>;
}

// ─── Types ──────────────────────────────────────────────

export interface EvolutionRuntimeConfig {
  /** Sync interval in ms (default: 60000 = 1 min) */
  syncIntervalMs?: number;
  /** Signal enrichment mode */
  enrichment?: SignalEnrichmentConfig;
  /** Scope for all operations (default: 'global') */
  scope?: string;
  /** Max outbox queue size before force flush */
  outboxMaxSize?: number;
  /** Outbox flush interval in ms (default: 5000) */
  outboxFlushMs?: number;
}

export interface Suggestion {
  action: 'apply_gene' | 'create_suggested' | 'none';
  geneId?: string;
  gene?: IMGene;
  strategy?: string[];
  confidence: number;
  signals: SignalTag[];
  fromCache: boolean;
  reason?: string;
  alternatives?: Array<{ gene_id: string; confidence: number; title?: string }>;
}

interface OutboxEntry {
  geneId: string;
  signals: SignalTag[];
  outcome: 'success' | 'failed';
  summary: string;
  score?: number;
  metadata?: Record<string, any>;
  timestamp: number;
  sessionId?: string;
}

/** Tracks a single suggest→learned cycle within a task. */
export interface EvolutionSession {
  /** Unique session ID */
  id: string;
  /** When suggest() was called */
  suggestedAt: number;
  /** What gene was recommended */
  suggestedGeneId?: string;
  /** What gene the agent actually used (may differ from suggested) */
  usedGeneId?: string;
  /** Signals that triggered the suggest */
  signals: SignalTag[];
  /** Whether agent adopted the recommended strategy */
  adopted: boolean;
  /** When learned() was called (null if not yet) */
  completedAt?: number;
  /** Outcome */
  outcome?: 'success' | 'failed';
  /** Duration from suggest to learned (ms) */
  durationMs?: number;
  /** Confidence of the suggestion */
  confidence: number;
  /** Whether suggestion came from cache or server */
  fromCache: boolean;
}

/** Aggregate session metrics for benchmarking. */
export interface SessionMetrics {
  /** Total suggest() calls */
  totalSuggestions: number;
  /** suggest() calls that returned a gene */
  suggestionsWithGene: number;
  /** learned() calls */
  totalLearned: number;
  /** Sessions where agent used the suggested gene */
  adoptedCount: number;
  /** Gene Utilization Rate = adoptedCount / suggestionsWithGene */
  geneUtilizationRate: number;
  /** Average suggest→learned duration (ms) */
  avgDurationMs: number;
  /** Success rate of adopted recommendations */
  adoptedSuccessRate: number;
  /** Success rate without adoption (agent did it alone) */
  nonAdoptedSuccessRate: number;
  /** Cache hit rate */
  cacheHitRate: number;
}

// ─── Runtime ────────────────────────────────────────────

export class EvolutionRuntime {
  private cache: EvolutionCache;
  private enricher: (ctx: ExecutionContext) => Promise<SignalTag[]>;
  private outbox: OutboxEntry[] = [];
  private syncTimer?: ReturnType<typeof setInterval>;
  private flushTimer?: ReturnType<typeof setInterval>;
  private lastSuggestedGeneId?: string;
  private started = false;
  private readonly scope: string;
  private readonly config: Required<EvolutionRuntimeConfig>;

  // Session tracking
  private _sessions: EvolutionSession[] = [];
  private _activeSession?: EvolutionSession;
  private _sessionCounter = 0;

  constructor(
    private client: EvolutionClientLike,
    config?: EvolutionRuntimeConfig,
  ) {
    this.config = {
      syncIntervalMs: config?.syncIntervalMs ?? 60_000,
      enrichment: config?.enrichment ?? { mode: 'rules' },
      scope: config?.scope ?? 'global',
      outboxMaxSize: config?.outboxMaxSize ?? 50,
      outboxFlushMs: config?.outboxFlushMs ?? 5_000,
    };
    this.scope = this.config.scope;
    this.cache = new EvolutionCache();
    this.enricher = config?.enrichment
      ? createEnrichedExtractor(config.enrichment)
      : async (ctx) => extractSignals(ctx);
  }

  // ─── Lifecycle ──────────────────────────────────────

  /** Initialize: load snapshot + start sync + start outbox flush */
  async start(): Promise<void> {
    if (this.started) return;
    this.started = true;

    // Bootstrap cache from server
    try {
      const snapshot = await this.client.getSyncSnapshot(0);
      if (snapshot.data) {
        this.cache.loadSnapshot(snapshot.data);
      }
    } catch {
      // Non-fatal: runtime works without cache (falls back to server)
    }

    // Periodic sync
    if (this.config.syncIntervalMs > 0) {
      this.syncTimer = setInterval(() => this.sync(), this.config.syncIntervalMs);
    }

    // Periodic outbox flush
    this.flushTimer = setInterval(() => this.flush(), this.config.outboxFlushMs);
  }

  /** Stop: clear timers + flush remaining outbox */
  async stop(): Promise<void> {
    if (this.syncTimer) clearInterval(this.syncTimer);
    if (this.flushTimer) clearInterval(this.flushTimer);
    await this.flush();
    this.started = false;
  }

  // ─── High-Level API ─────────────────────────────────

  /**
   * Get a strategy recommendation for an error/context.
   *
   * Flow: extract signals → try local cache (<1ms) → fallback to server (~30ms)
   *
   * @param error - Error message or Error object
   * @param context - Optional additional context (provider, stage, etc.)
   */
  async suggest(
    error: string | Error,
    context?: Partial<ExecutionContext>,
  ): Promise<Suggestion> {
    const errorStr = error instanceof Error ? error.message : error;
    const ctx: ExecutionContext = {
      error: errorStr,
      ...context,
    };

    // Extract signals
    const signals = await this.enricher(ctx);
    if (signals.length === 0) {
      return {
        action: 'none',
        confidence: 0,
        signals: [],
        fromCache: false,
        reason: 'no signals extracted from error',
      };
    }

    const buildSuggestion = (
      action: string, geneId: string | undefined, gene: any, strategy: any,
      confidence: number, fromCache: boolean, reason?: string, alternatives?: any,
    ): Suggestion => {
      this.lastSuggestedGeneId = geneId;
      // Start session tracking
      this._activeSession = {
        id: `ses_${++this._sessionCounter}_${Date.now()}`,
        suggestedAt: Date.now(),
        suggestedGeneId: geneId,
        signals,
        adopted: false,
        confidence,
        fromCache,
      };
      return {
        action: action as any, geneId, gene, strategy, confidence,
        signals, fromCache, reason, alternatives,
      };
    };

    // Try local cache first (<1ms)
    if (this.cache.geneCount > 0) {
      const local = this.cache.selectGene(signals);
      if (local.action === 'apply_gene' && local.confidence > 0.3) {
        return buildSuggestion(
          local.action, local.gene_id, local.gene, local.strategy,
          local.confidence, true, local.reason, local.alternatives,
        );
      }
    }

    // Fallback to server
    try {
      const result = await this.client.analyze({
        signals: signals as any,
        scope: this.scope,
      });
      if (result.data) {
        return buildSuggestion(
          result.data.action, result.data.gene_id, result.data.gene, result.data.strategy,
          result.data.confidence ?? 0, false, result.data.reason, result.data.alternatives,
        );
      }
    } catch {
      const local = this.cache.selectGene(signals);
      return buildSuggestion(
        local.action, local.gene_id, local.gene, local.strategy,
        local.confidence, true, 'server unreachable, using cache fallback', local.alternatives,
      );
    }

    return {
      action: 'none',
      confidence: 0,
      signals,
      fromCache: false,
      reason: 'no recommendation from server',
    };
  }

  /**
   * Record an outcome. Fire-and-forget — never blocks, never throws.
   *
   * @param error - The error that was encountered
   * @param outcome - 'success' or 'failed'
   * @param summary - One-line summary of what happened
   * @param geneId - Gene that was used (auto-detected from last suggest() if omitted)
   */
  learned(
    error: string | Error,
    outcome: 'success' | 'failed',
    summary: string,
    geneId?: string,
    metadata?: Record<string, any>,
  ): void {
    const errorStr = error instanceof Error ? error.message : error;
    const ctx: ExecutionContext = { error: errorStr };
    const signals = extractSignals(ctx); // Sync extraction for non-blocking

    const resolvedGeneId = geneId || this.lastSuggestedGeneId;
    if (!resolvedGeneId) return; // Can't record without a gene

    // Complete active session
    if (this._activeSession) {
      const session = this._activeSession;
      session.usedGeneId = resolvedGeneId;
      session.adopted = resolvedGeneId === session.suggestedGeneId;
      session.completedAt = Date.now();
      session.outcome = outcome;
      session.durationMs = session.completedAt - session.suggestedAt;
      this._sessions.push(session);
      this._activeSession = undefined;
    }

    this.outbox.push({
      geneId: resolvedGeneId,
      signals,
      outcome,
      summary,
      metadata,
      timestamp: Date.now(),
      sessionId: this._sessions[this._sessions.length - 1]?.id,
    });

    // Force flush if outbox is full
    if (this.outbox.length >= this.config.outboxMaxSize) {
      this.flush().catch(() => {});
    }
  }

  // ─── Session Metrics ────────────────────────────────

  /** Get all completed sessions. */
  get sessions(): readonly EvolutionSession[] {
    return this._sessions;
  }

  /** Get aggregate metrics for benchmarking. */
  getMetrics(): SessionMetrics {
    const sessions = this._sessions;
    const totalSuggestions = sessions.length;
    const suggestionsWithGene = sessions.filter(s => s.suggestedGeneId).length;
    const totalLearned = sessions.filter(s => s.completedAt).length;
    const adoptedSessions = sessions.filter(s => s.adopted && s.completedAt);
    const adoptedCount = adoptedSessions.length;
    const nonAdopted = sessions.filter(s => !s.adopted && s.completedAt);

    const durations = sessions.filter(s => s.durationMs != null).map(s => s.durationMs!);
    const avgDurationMs = durations.length > 0
      ? Math.round(durations.reduce((a, b) => a + b, 0) / durations.length)
      : 0;

    const adoptedSuccess = adoptedSessions.filter(s => s.outcome === 'success').length;
    const nonAdoptedSuccess = nonAdopted.filter(s => s.outcome === 'success').length;
    const cacheHits = sessions.filter(s => s.fromCache).length;

    return {
      totalSuggestions,
      suggestionsWithGene,
      totalLearned,
      adoptedCount,
      geneUtilizationRate: suggestionsWithGene > 0 ? Math.round(adoptedCount / suggestionsWithGene * 100) / 100 : 0,
      avgDurationMs,
      adoptedSuccessRate: adoptedCount > 0 ? Math.round(adoptedSuccess / adoptedCount * 100) / 100 : 0,
      nonAdoptedSuccessRate: nonAdopted.length > 0 ? Math.round(nonAdoptedSuccess / nonAdopted.length * 100) / 100 : 0,
      cacheHitRate: totalSuggestions > 0 ? Math.round(cacheHits / totalSuggestions * 100) / 100 : 0,
    };
  }

  /** Reset session history. */
  resetMetrics(): void {
    this._sessions = [];
  }

  // ─── Internal ───────────────────────────────────────

  /** Sync cache with server */
  private async sync(): Promise<void> {
    try {
      const result = await this.client.sync({
        pull: { since: this.cache.cursor },
        scope: this.scope,
      } as any);
      if (result.data?.pulled) {
        this.cache.applyDelta({ pulled: result.data.pulled });
      }
    } catch {
      // Non-fatal
    }
  }

  /** Flush outbox to server */
  private async flush(): Promise<void> {
    if (this.outbox.length === 0) return;

    const batch = this.outbox.splice(0, this.config.outboxMaxSize);
    const promises = batch.map(entry =>
      this.client.record({
        gene_id: entry.geneId,
        signals: entry.signals.map(s => s.type),
        outcome: entry.outcome,
        summary: entry.summary,
        score: entry.score,
        metadata: entry.metadata,
        scope: this.scope,
      } as any).catch(() => {
        // Put back on failure
        this.outbox.push(entry);
      })
    );

    await Promise.allSettled(promises);
  }
}

File: sdk/python/prismer/evolution_runtime.py
"""EvolutionRuntime — High-level evolution API for Python agents.

Composes EvolutionCache + SignalEnrichment + async outbox into two simple methods:
  - suggest(error, context?) → strategy recommendation (<1ms local, fallback to server)
  - learned(error, outcome, summary, gene_id?) → fire-and-forget outcome recording

Port of sdk/typescript/src/evolution-runtime.ts.

Usage:
    from prismer import PrismerClient
    from prismer.evolution_runtime import EvolutionRuntime

    client = PrismerClient(api_key="sk-...")
    runtime = EvolutionRuntime(client.im.evolution)
    await runtime.start()

    fix = await runtime.suggest("ETIMEDOUT: connection timed out")
    # ... agent applies fix.strategy ...
    runtime.learned("ETIMEDOUT", "success", "Fixed by increasing timeout")
"""

from __future__ import annotations

import asyncio
import threading
import time
from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List, Optional, Protocol

from .evolution_cache import EvolutionCache, GeneSelectionResult, SignalTag
from .signal_rules import extract_signals


# ─── Protocols ──────────────────────────────────────────

class EvolutionClientLike(Protocol):
    """Minimal interface — works with both sync and async EvolutionClient."""

    def get_sync_snapshot(self, since: int = 0) -> Any: ...
    def analyze(self, **kwargs: Any) -> Any: ...
    def record(self, gene_id: str, signals: list, outcome: str, summary: str, **kwargs: Any) -> Any: ...
    def sync(self, push: Any = None, pull: Any = None) -> Any: ...


class AsyncEvolutionClientLike(Protocol):
    async def get_sync_snapshot(self, since: int = 0) -> Any: ...
    async def analyze(self, **kwargs: Any) -> Any: ...
    async def record(self, gene_id: str, signals: list, outcome: str, summary: str, **kwargs: Any) -> Any: ...
    async def sync(self, push: Any = None, pull: Any = None) -> Any: ...


# ─── Types ──────────────────────────────────────────────

@dataclass
class EvolutionRuntimeConfig:
    sync_interval_s: float = 60.0
    scope: str = "global"
    outbox_max_size: int = 50
    outbox_flush_s: float = 5.0


@dataclass
class Suggestion:
    action: str  # 'apply_gene' | 'create_suggested' | 'none'
    gene_id: Optional[str] = None
    gene: Optional[Dict[str, Any]] = None
    strategy: Optional[List[str]] = None
    confidence: float = 0.0
    signals: List[Dict[str, str]] = field(default_factory=list)
    from_cache: bool = False
    reason: Optional[str] = None
    alternatives: Optional[List[Dict[str, Any]]] = None


@dataclass
class EvolutionSession:
    """Tracks a single suggest→learned cycle."""
    id: str
    suggested_at: float
    suggested_gene_id: Optional[str] = None
    used_gene_id: Optional[str] = None
    signals: List[Dict[str, str]] = field(default_factory=list)
    adopted: bool = False
    completed_at: Optional[float] = None
    outcome: Optional[str] = None
    duration_ms: Optional[float] = None
    confidence: float = 0.0
    from_cache: bool = False


@dataclass
class SessionMetrics:
    """Aggregate session metrics for benchmarking."""
    total_suggestions: int = 0
    suggestions_with_gene: int = 0
    total_learned: int = 0
    adopted_count: int = 0
    gene_utilization_rate: float = 0.0
    avg_duration_ms: float = 0.0
    adopted_success_rate: float = 0.0
    non_adopted_success_rate: float = 0.0
    cache_hit_rate: float = 0.0


@dataclass
class _OutboxEntry:
    gene_id: str
    signals: List[Dict[str, str]]
    outcome: str
    summary: str
    score: Optional[float] = None
    metadata: Optional[Dict[str, Any]] = None
    timestamp: float = 0.0


# ─── Async Runtime ──────────────────────────────────────

class AsyncEvolutionRuntime:
    """Async evolution runtime for asyncio-based agents."""

    def __init__(
        self,
        client: AsyncEvolutionClientLike,
        config: Optional[EvolutionRuntimeConfig] = None,
    ) -> None:
        self._client = client
        self._config = config or EvolutionRuntimeConfig()
        self._cache = EvolutionCache()
        self._outbox: List[_OutboxEntry] = []
        self._last_suggested_gene_id: Optional[str] = None
        self._started = False
        self._sync_task: Optional[asyncio.Task[None]] = None
        self._flush_task: Optional[asyncio.Task[None]] = None
        self._sessions: List[EvolutionSession] = []
        self._active_session: Optional[EvolutionSession] = None
        self._session_counter = 0

    async def start(self) -> None:
        """Bootstrap: load snapshot + start sync + start flush."""
        if self._started:
            return
        self._started = True

        try:
            result = self._client.get_sync_snapshot(0)
            if asyncio.iscoroutine(result):
                result = await result
            data = getattr(result, "data", None) or (result if isinstance(result, dict) else None)
            if isinstance(data, dict) and "data" in data:
                data = data["data"]
            if data:
                self._cache.load_snapshot(data)
        except Exception:
            pass

        if self._config.sync_interval_s > 0:
            self._sync_task = asyncio.create_task(self._sync_loop())
        self._flush_task = asyncio.create_task(self._flush_loop())

    async def stop(self) -> None:
        """Stop timers + flush remaining outbox."""
        self._started = False
        if self._sync_task:
            self._sync_task.cancel()
        if self._flush_task:
            self._flush_task.cancel()
        await self._flush()

    async def suggest(
        self,
        error: str,
        *,
        provider: Optional[str] = None,
        stage: Optional[str] = None,
        severity: Optional[str] = None,
        tags: Optional[List[str]] = None,
    ) -> Suggestion:
        """Get a strategy recommendation. Cache first (<1ms), server fallback."""
        signals = extract_signals(
            error=error, provider=provider, stage=stage, severity=severity, tags=tags,
        )
        if not signals:
            return Suggestion(action="none", reason="no signals extracted from error")

        signal_tags = [SignalTag(type=s["type"]) for s in signals]

        # Try local cache
        if self._cache.gene_count > 0:
            local = self._cache.select_gene(signal_tags)
            if local.action == "apply_gene" and local.confidence > 0.3:
                self._last_suggested_gene_id = local.gene_id
                self._start_session(local.gene_id, signals, local.confidence, True)
                return Suggestion(
                    action=local.action,
                    gene_id=local.gene_id,
                    gene=local.gene,
                    strategy=local.strategy,
                    confidence=local.confidence,
                    signals=signals,
                    from_cache=True,
                    reason=local.reason,
                    alternatives=local.alternatives,
                )

        # Fallback to server
        try:
            result = self._client.analyze(signals=signals, scope=self._config.scope)
            if asyncio.iscoroutine(result):
                result = await result
            data = getattr(result, "data", None)
            if isinstance(result, dict):
                data = result.get("data", result)
            if data:
                self._last_suggested_gene_id = data.get("gene_id")
                self._start_session(data.get("gene_id"), signals, data.get("confidence", 0), False)
                return Suggestion(
                    action=data.get("action", "none"),
                    gene_id=data.get("gene_id"),
                    gene=data.get("gene"),
                    strategy=data.get("strategy"),
                    confidence=data.get("confidence", 0),
                    signals=signals,
                    from_cache=False,
                    reason=data.get("reason"),
                    alternatives=data.get("alternatives"),
                )
        except Exception:
            # Server unreachable — use cache even if low confidence
            local = self._cache.select_gene(signal_tags)
            self._last_suggested_gene_id = local.gene_id
            self._start_session(local.gene_id, signals, local.confidence, True)
            return Suggestion(
                action=local.action,
                gene_id=local.gene_id,
                gene=local.gene,
                strategy=local.strategy,
                confidence=local.confidence,
                signals=signals,
                from_cache=True,
                reason="server unreachable, using cache fallback",
            )

        return Suggestion(action="none", signals=signals, reason="no recommendation")

    def _start_session(self, gene_id: Optional[str], signals: List[Dict[str, str]], confidence: float, from_cache: bool) -> None:
        self._session_counter += 1
        self._active_session = EvolutionSession(
            id=f"ses_{self._session_counter}_{int(time.time()*1000)}",
            suggested_at=time.time(),
            suggested_gene_id=gene_id,
            signals=signals,
            confidence=confidence,
            from_cache=from_cache,
        )

    def _complete_session(self, gene_id: str, outcome: str) -> None:
        if self._active_session:
            s = self._active_session
            s.used_gene_id = gene_id
            s.adopted = gene_id == s.suggested_gene_id
            s.completed_at = time.time()
            s.outcome = outcome
            s.duration_ms = (s.completed_at - s.suggested_at) * 1000
            self._sessions.append(s)
            self._active_session = None

    @property
    def sessions(self) -> List[EvolutionSession]:
        return list(self._sessions)

    def get_metrics(self) -> SessionMetrics:
        ss = self._sessions
        total = len(ss)
        with_gene = sum(1 for s in ss if s.suggested_gene_id)
        learned = sum(1 for s in ss if s.completed_at)
        adopted = [s for s in ss if s.adopted and s.completed_at]
        non_adopted = [s for s in ss if not s.adopted and s.completed_at]
        durations = [s.duration_ms for s in ss if s.duration_ms is not None]
        cache_hits = sum(1 for s in ss if s.from_cache)

        return SessionMetrics(
            total_suggestions=total,
            suggestions_with_gene=with_gene,
            total_learned=learned,
            adopted_count=len(adopted),
            gene_utilization_rate=round(len(adopted) / with_gene, 2) if with_gene else 0,
            avg_duration_ms=round(sum(durations) / len(durations)) if durations else 0,
            adopted_success_rate=round(sum(1 for s in adopted if s.outcome == "success") / len(adopted), 2) if adopted else 0,
            non_adopted_success_rate=round(sum(1 for s in non_adopted if s.outcome == "success") / len(non_adopted), 2) if non_adopted else 0,
            cache_hit_rate=round(cache_hits / total, 2) if total else 0,
        )

    def reset_metrics(self) -> None:
        """Clear all sessions and reset counter."""
        self._sessions.clear()
        self._active_session = None
        self._session_counter = 0

    def learned(
        self,
        error: str,
        outcome: str,
        summary: str,
        gene_id: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> None:
        """Record outcome. Fire-and-forget — never blocks, never throws."""
        signals = extract_signals(error=error)
        resolved = gene_id or self._last_suggested_gene_id
        if not resolved:
            return

        self._complete_session(resolved, outcome)

        self._outbox.append(_OutboxEntry(
            gene_id=resolved,
            signals=signals,
            outcome=outcome,
            summary=summary,
            metadata=metadata,
            timestamp=time.time(),
        ))

        if len(self._outbox) >= self._config.outbox_max_size:
            asyncio.ensure_future(self._flush())

    # ─── Internal ───────────────────────────────────────

    async def _sync_loop(self) -> None:
        while self._started:
            await asyncio.sleep(self._config.sync_interval_s)
            try:
                result = self._client.sync(pull={"since": self._cache.cursor})
                if asyncio.iscoroutine(result):
                    result = await result
                data = getattr(result, "data", None) or (result if isinstance(result, dict) else None)
                if isinstance(data, dict) and "pulled" in data:
                    self._cache.apply_delta({"pulled": data["pulled"]})
            except Exception:
                pass

    async def _flush_loop(self) -> None:
        while self._started:
            await asyncio.sleep(self._config.outbox_flush_s)
            await self._flush()

    async def _flush(self) -> None:
        if not self._outbox:
            return
        batch = self._outbox[: self._config.outbox_max_size]
        self._outbox = self._outbox[self._config.outbox_max_size :]

        for entry in batch:
            try:
                result = self._client.record(
                    gene_id=entry.gene_id,
                    signals=[s["type"] for s in entry.signals],
                    outcome=entry.outcome,
                    summary=entry.summary,
                    scope=self._config.scope,
                )
                if asyncio.iscoroutine(result):
                    await result
            except Exception:
                self._outbox.append(entry)


# ─── Sync Runtime (thread-based) ────────────────────────

class EvolutionRuntime:
    """Sync evolution runtime for thread-based agents.

    Wraps AsyncEvolutionRuntime via threading for agents that don't use asyncio.
    """

    def __init__(
        self,
        client: EvolutionClientLike,
        config: Optional[EvolutionRuntimeConfig] = None,
    ) -> None:
        self._client = client
        self._config = config or EvolutionRuntimeConfig()
        self._cache = EvolutionCache()
        self._outbox: List[_OutboxEntry] = []
        self._last_suggested_gene_id: Optional[str] = None
        self._started = False
        self._lock = threading.Lock()
        self._flush_timer: Optional[threading.Timer] = None
        self._sessions: List[EvolutionSession] = []
        self._active_session: Optional[EvolutionSession] = None
        self._session_counter = 0

    def start(self) -> None:
        """Bootstrap: load snapshot synchronously."""
        if self._started:
            return
        self._started = True

        try:
            result = self._client.get_sync_snapshot(0)
            data = getattr(result, "data", None) or (result if isinstance(result, dict) else None)
            if isinstance(data, dict) and "data" in data:
                data = data["data"]
            if data:
                self._cache.load_snapshot(data)
        except Exception:
            pass

        self._schedule_flush()

    def stop(self) -> None:
        self._started = False
        if self._flush_timer:
            self._flush_timer.cancel()
        self._flush()

    def suggest(
        self,
        error: str,
        *,
        provider: Optional[str] = None,
        stage: Optional[str] = None,
        severity: Optional[str] = None,
        tags: Optional[List[str]] = None,
    ) -> Suggestion:
        """Get a strategy recommendation. Cache first, server fallback."""
        signals = extract_signals(error=error, provider=provider, stage=stage, severity=severity, tags=tags)
        if not signals:
            return Suggestion(action="none", reason="no signals extracted")

        signal_tags = [SignalTag(type=s["type"]) for s in signals]

        if self._cache.gene_count > 0:
            local = self._cache.select_gene(signal_tags)
            if local.action == "apply_gene" and local.confidence > 0.3:
                self._last_suggested_gene_id = local.gene_id
                self._start_session(local.gene_id, signals, local.confidence, True)
                return Suggestion(
                    action=local.action, gene_id=local.gene_id, gene=local.gene,
                    strategy=local.strategy, confidence=local.confidence,
                    signals=signals, from_cache=True, reason=local.reason,
                )

        try:
            result = self._client.analyze(signals=signals, scope=self._config.scope)
            data = getattr(result, "data", None) or result
            if isinstance(data, dict):
                self._last_suggested_gene_id = data.get("gene_id")
                self._start_session(data.get("gene_id"), signals, data.get("confidence", 0), False)
                return Suggestion(
                    action=data.get("action", "none"), gene_id=data.get("gene_id"),
                    gene=data.get("gene"), strategy=data.get("strategy"),
                    confidence=data.get("confidence", 0), signals=signals,
                    from_cache=False, reason=data.get("reason"),
                )
        except Exception:
            # Server unreachable — use cache even if low confidence
            local = self._cache.select_gene(signal_tags)
            self._last_suggested_gene_id = local.gene_id
            self._start_session(local.gene_id, signals, local.confidence, True)
            return Suggestion(
                action=local.action, gene_id=local.gene_id, gene=local.gene,
                strategy=local.strategy, confidence=local.confidence,
                signals=signals, from_cache=True, reason="server unreachable, using cache fallback",
            )

        return Suggestion(action="none", signals=signals, reason="no recommendation")

    def learned(
        self,
        error: str,
        outcome: str,
        summary: str,
        gene_id: Optional[str] = None,
    ) -> None:
        """Record outcome. Fire-and-forget."""
        signals = extract_signals(error=error)
        resolved = gene_id or self._last_suggested_gene_id
        if not resolved:
            return

        self._complete_session(resolved, outcome)

        with self._lock:
            self._outbox.append(_OutboxEntry(
                gene_id=resolved, signals=signals, outcome=outcome,
                summary=summary, timestamp=time.time(),
            ))

        if len(self._outbox) >= self._config.outbox_max_size:
            self._flush()

    # ─── Session tracking ─────────────────────────────────

    def _start_session(self, gene_id: Optional[str], signals: List[Dict[str, str]], confidence: float, from_cache: bool) -> None:
        if self._active_session:
            self._sessions.append(self._active_session)
        self._session_counter += 1
        self._active_session = EvolutionSession(
            id=f"ses_{self._session_counter}_{int(time.time()*1000)}",
            suggested_at=time.time(),
            suggested_gene_id=gene_id,
            signals=signals,
            confidence=confidence,
            from_cache=from_cache,
        )

    def _complete_session(self, gene_id: str, outcome: str) -> None:
        if self._active_session:
            s = self._active_session
            s.used_gene_id = gene_id
            s.adopted = gene_id == s.suggested_gene_id
            s.completed_at = time.time()
            s.outcome = outcome
            s.duration_ms = (s.completed_at - s.suggested_at) * 1000
            self._sessions.append(s)
            self._active_session = None

    @property
    def sessions(self) -> List[EvolutionSession]:
        return list(self._sessions)

    def get_metrics(self) -> SessionMetrics:
        ss = self._sessions
        total = len(ss)
        with_gene = sum(1 for s in ss if s.suggested_gene_id)
        learned_count = sum(1 for s in ss if s.completed_at)
        adopted = [s for s in ss if s.adopted and s.completed_at]
        non_adopted = [s for s in ss if not s.adopted and s.completed_at]
        durations = [s.duration_ms for s in ss if s.duration_ms is not None]
        cache_hits = sum(1 for s in ss if s.from_cache)

        return SessionMetrics(
            total_suggestions=total,
            suggestions_with_gene=with_gene,
            total_learned=learned_count,
            adopted_count=len(adopted),
            gene_utilization_rate=round(len(adopted) / with_gene, 2) if with_gene else 0,
            avg_duration_ms=round(sum(durations) / len(durations)) if durations else 0,
            adopted_success_rate=round(sum(1 for s in adopted if s.outcome == "success") / len(adopted), 2) if adopted else 0,
            non_adopted_success_rate=round(sum(1 for s in non_adopted if s.outcome == "success") / len(non_adopted), 2) if non_adopted else 0,
            cache_hit_rate=round(cache_hits / total, 2) if total else 0,
        )

    def reset_metrics(self) -> None:
        """Clear all sessions and reset counter."""
        self._sessions.clear()
        self._active_session = None
        self._session_counter = 0

    def _flush(self) -> None:
        with self._lock:
            batch = self._outbox[:]
            self._outbox.clear()

        for entry in batch:
            try:
                self._client.record(
                    gene_id=entry.gene_id,
                    signals=[s["type"] for s in entry.signals],
                    outcome=entry.outcome,
                    summary=entry.summary,
                    scope=self._config.scope,
                )
            except Exception:
                with self._lock:
                    self._outbox.append(entry)

    def _schedule_flush(self) -> None:
        if not self._started:
            return
        self._flush_timer = threading.Timer(self._config.outbox_flush_s, self._do_flush_and_reschedule)
        self._flush_timer.daemon = True
        self._flush_timer.start()

    def _do_flush_and_reschedule(self) -> None:
        self._flush()
        self._schedule_flush()
